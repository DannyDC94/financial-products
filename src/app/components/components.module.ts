import { NgModule } from '@angular/core';
import {CommonModule, NgOptimizedImage} from '@angular/common';
import { HeaderComponent } from './header/header.component';
import { DataTableComponent } from './data-table/data-table.component'
import {SharedModule} from "../shared/shared.module";
import {FormsModule, ReactiveFormsModule} from "@angular/forms";
import { ModalComponent } from './modal/modal.component';
import { FormDataComponent } from './form-data/form-data.component';
@NgModule({
    exports: [HeaderComponent, DataTableComponent, ModalComponent, FormDataComponent],
  declarations: [HeaderComponent, DataTableComponent, ModalComponent, FormDataComponent],
  imports: [
    CommonModule,
    SharedModule,
    FormsModule,
    ReactiveFormsModule,
    NgOptimizedImage
  ]
})
export class ComponentsModule { }
