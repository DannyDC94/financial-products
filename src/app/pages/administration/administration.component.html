<app-header [logo]="imgHeader"></app-header>

<main>
    <div class="container-skeleton" *ngIf="!showTable; else loadedContent">
      <ngx-skeleton-loader count="5" [theme]="skeletonTheme"></ngx-skeleton-loader>
    </div>
    <ng-template #loadedContent>
      <app-data-table [tableData]="dataTable" [tableOptions]="tableOptions" (btnActions)="actionProduct($event)"></app-data-table>
    </ng-template>
</main>

<ngx-spinner type="ball-scale-multiple"><p>Cargando...</p></ngx-spinner>

<app-modal *ngIf="showConfirmModal" [message]="confirmMessage" (confirm)="onConfirm()" (cancel)="onCancel()"></app-modal>

<app-modal *ngIf="showModalMessage" [type]="'alert'" [message]="modalMessage" (confirm)="onCancel()"></app-modal>

