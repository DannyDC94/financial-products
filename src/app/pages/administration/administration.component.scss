.container-skeleton {
  margin: 5rem 8rem;
}

p {
  color: #ffffff;
}
