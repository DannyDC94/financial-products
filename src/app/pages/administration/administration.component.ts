import { Component, OnDestroy, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { NgxSpinnerService } from "ngx-spinner";
import { Subscription } from "rxjs";

import { FinancialProductsService } from '../../services/financial-products.service';
import { Product } from "../../interfaces";

const SAVE_PRODUCT_ROUTE = '/save-product';
const EDIT_ACTION = 'edit';
const DELETE_ACTION = 'delete';

@Component({
  selector: 'app-administration',
  templateUrl: './administration.component.html',
  styleUrls: ['./administration.component.scss']
})

export class AdministrationComponent implements OnInit, OnDestroy {

  constructor(
    private financialSvc: FinancialProductsService,
    private router: Router,
    private spinner: NgxSpinnerService
  ) {
  }

  imgHeader: string = 'assets/imgs/BP_logo.svg'
  tableOptions: any;
  dataTable: Product[] = [];
  showTable = false;
  showConfirmModal = false;
  idSelect: string = '';
  nombreSelect: string = '';
  confirmMessage = '';
  showModalMessage = false;
  modalMessage: string = '';
  skeletonTheme: object = {
    'border-radius': '5px',
    'height': '50px',
    'background-color': '#e5e5e5',
    'border': '1px solid white'
  };
  private subscriptions: Subscription[] = [];

  ngOnInit() {
    this.loadColumnsTable();
    this.loadProducts();
  }

  loadColumnsTable() {
    const columns = [
      {field: 'logo', name: 'Logo', type: 'image', breakpoint: 'sm'},
      {field: 'name', name: 'Nombre del producto', breakpoint: ''},
      {field: 'description', name: 'Descripción', breakpoint: 'lg'},
      {field: 'date_release', name: 'Fecha de liberación', type: 'date', breakpoint: 'lg'},
      {field: 'date_revision', name: 'Fecha de reestructuración', type: 'date', breakpoint: 'md'},
      {field: 'isDropdownActive', name: '', type: 'actions', breakpoint: 'actions'}
    ]
    this.tableOptions = {
      columns: columns,
    };
  }

  loadProducts() {
    this.showTable = false;
    this.subscriptions.push(
      this.financialSvc.getFinancialProducts().subscribe({
        next: (v: Product[]) => this.dataTable = v,
        error: () => this.openModalMessage('Existen errores al cargar productos'),
        complete: () => this.showTable = true
      })
    );
  }

  actionProduct(evt: any) {
    const { action, data } = evt;
    if (action === EDIT_ACTION) {
      this.router.navigate([SAVE_PRODUCT_ROUTE], {
        queryParams: {
          type: action,
          data: JSON.stringify(data)
        }
      });
    } else if (action === DELETE_ACTION) {
      this.idSelect = data.id;
      this.nombreSelect = data.name;
      this.openConfirmModal();
    } else {
      this.router.navigate([SAVE_PRODUCT_ROUTE], {
        queryParams: {
          type: action
        }
      });
    }
  }

  deleteProduct(id: string) {
    this.spinner.show();
    this.subscriptions.push(this.financialSvc.deleteFinancialProducts(id).subscribe({
      next: (v) => {
        if (v.text.indexOf('Product successfully removed') > -1) {
          this.spinner.hide()
          this.loadProducts();
        } else {
          this.openModalMessage('Existen errores el eliminar el producto');
        }
      },
      error: (e) => {
        this.openModalMessage('Existen errores el eliminar el producto');
        this.spinner.hide();
      },
      complete: () => this.spinner.hide()
    }));
  }
  openConfirmModal(): void {
    this.confirmMessage = `¿Estas seguro de eliminar el producto ${this.nombreSelect}?`
    this.showConfirmModal = true;
  }

  onConfirm(): void {
    this.showConfirmModal = false;
    this.deleteProduct(this.idSelect);
  }

  onCancel(): void {
    this.showConfirmModal = false;
    this.showModalMessage = false;
  }

  openModalMessage(message: string): void {
    this.showModalMessage = true;
    this.modalMessage = message;
  }

  ngOnDestroy() {
    this.subscriptions.forEach(sb => sb.unsubscribe());
  }

}
