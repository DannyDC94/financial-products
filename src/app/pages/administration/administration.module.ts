import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { AdministrationRoutingModule } from './administration-routing.module';
import { AdministrationComponent } from './administration.component';
import { ComponentsModule } from '../../components/components.module';
import { FinancialProductsService } from '../../services/financial-products.service'
import {NgxSkeletonLoaderModule} from "ngx-skeleton-loader";
import {NgxSpinnerModule} from "ngx-spinner";


@NgModule({
  declarations: [
    AdministrationComponent
  ],
    imports: [
        CommonModule,
        AdministrationRoutingModule,
        ComponentsModule,
        NgxSkeletonLoaderModule,
        NgxSpinnerModule
    ],
  providers: [FinancialProductsService]
})
export class AdministrationModule { }
