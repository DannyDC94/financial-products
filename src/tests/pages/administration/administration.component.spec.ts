import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpTestingController, HttpClientTestingModule} from '@angular/common/http/testing'
import { AdministrationComponent } from '../../../app/pages/administration/administration.component';
import { FinancialProductsService } from '../../../app/services/financial-products.service';
import { environment } from '../../../environment/environment';
import { NgxSkeletonLoaderModule } from "ngx-skeleton-loader";
import { ComponentsModule } from "../../../app/components/components.module";
import {NgxSpinnerModule} from "ngx-spinner";

describe('AdministrationComponent', () => {
  let component: AdministrationComponent;
  let fixture: ComponentFixture<AdministrationComponent>;
  let compile:  HTMLElement;
  let service: FinancialProductsService;
  let httpMock: HttpTestingController;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [AdministrationComponent],
      imports: [
        HttpClientTestingModule,
        NgxSkeletonLoaderModule,
        ComponentsModule,
        NgxSpinnerModule
      ],
      providers: [FinancialProductsService]
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(AdministrationComponent);
    component = fixture.componentInstance;
    service = TestBed.inject(FinancialProductsService);
    httpMock = TestBed.inject(HttpTestingController)

    fixture.detectChanges();
    compile = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('Debe hacer match con el snapshot', () => {
    expect(compile.innerHTML).toMatchSnapshot();
  });

  test('Cargar datos de productos', () => {
    const dummyProducts = [
      {
        id: "trg-111",
        name: "Tarjetas",
        description: "Tarjetas para compras exterior",
        logo: "https://upload.wikimedia.org/wikipedia/commons/5/5e/Visa_Inc._logo.svg",
        date_release: "2023-12-04T00:00:00.000+00:00",
        date_revision: "2024-12-04T00:00:00.000+00:00"
      }
    ];
    const request = httpMock.expectOne(environment.apiUrl);
    expect(request.request.method).toBe('GET');
    request.flush(dummyProducts);
    fixture.detectChanges();
  })

  // test('Abrir modal confirm', () => {
  //   component.openConfirmModal();
  //   component.confirmMessage = 'tg-1';
  //   expect(component.confirmMessage).toEqual('Â¿Estas seguro de eliminar el producto tg-1?');
  // });
});
