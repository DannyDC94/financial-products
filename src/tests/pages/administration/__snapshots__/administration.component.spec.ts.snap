// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdministrationComponent Debe hacer match con el snapshot 1`] = `
"<app-header ng-reflect-logo="assets/imgs/BP_logo.svg"><header><img alt="Pichincha" src="assets/imgs/BP_logo.svg"></header></app-header><main><div class="container-skeleton"><ngx-skeleton-loader count="5" _nghost-a-c3247267950="" ng-reflect-count="5" ng-reflect-theme="[object Object]"><div _ngcontent-a-c3247267950="" aria-busy="true" aria-valuemin="0" aria-valuemax="100" role="progressbar" tabindex="-1" class="skeleton-loader progress" ng-reflect-ng-class="[object Object]" ng-reflect-ng-style="[object Object]" aria-label="loading" aria-valuetext="Loading..." style="border-radius: 5px; height: 50px; background-color: rgb(229, 229, 229); border: 1px solid white;"><!--bindings={
  "ng-reflect-ng-if": "false"
}--></div><div _ngcontent-a-c3247267950="" aria-busy="true" aria-valuemin="0" aria-valuemax="100" role="progressbar" tabindex="-1" class="skeleton-loader progress" ng-reflect-ng-class="[object Object]" ng-reflect-ng-style="[object Object]" aria-label="loading" aria-valuetext="Loading..." style="border-radius: 5px; height: 50px; background-color: rgb(229, 229, 229); border: 1px solid white;"><!--bindings={
  "ng-reflect-ng-if": "false"
}--></div><div _ngcontent-a-c3247267950="" aria-busy="true" aria-valuemin="0" aria-valuemax="100" role="progressbar" tabindex="-1" class="skeleton-loader progress" ng-reflect-ng-class="[object Object]" ng-reflect-ng-style="[object Object]" aria-label="loading" aria-valuetext="Loading..." style="border-radius: 5px; height: 50px; background-color: rgb(229, 229, 229); border: 1px solid white;"><!--bindings={
  "ng-reflect-ng-if": "false"
}--></div><div _ngcontent-a-c3247267950="" aria-busy="true" aria-valuemin="0" aria-valuemax="100" role="progressbar" tabindex="-1" class="skeleton-loader progress" ng-reflect-ng-class="[object Object]" ng-reflect-ng-style="[object Object]" aria-label="loading" aria-valuetext="Loading..." style="border-radius: 5px; height: 50px; background-color: rgb(229, 229, 229); border: 1px solid white;"><!--bindings={
  "ng-reflect-ng-if": "false"
}--></div><div _ngcontent-a-c3247267950="" aria-busy="true" aria-valuemin="0" aria-valuemax="100" role="progressbar" tabindex="-1" class="skeleton-loader progress" ng-reflect-ng-class="[object Object]" ng-reflect-ng-style="[object Object]" aria-label="loading" aria-valuetext="Loading..." style="border-radius: 5px; height: 50px; background-color: rgb(229, 229, 229); border: 1px solid white;"><!--bindings={
  "ng-reflect-ng-if": "false"
}--></div><!--bindings={
  "ng-reflect-ng-for-of": ",,,,"
}--></ngx-skeleton-loader></div><!--bindings={
  "ng-reflect-ng-if-else": "[object Object]"
}--><!--container--></main><ngx-spinner type="ball-scale-multiple" _nghost-a-c2009170884="" ng-reflect-type="ball-scale-multiple"><!--bindings={
  "ng-reflect-ng-if": "false"
}--></ngx-spinner><!--bindings={
  "ng-reflect-ng-if": "false"
}-->"
`;
